### PR TITLE
Merge Wip/headnode bootstrap error branch to develop

### PR DESCRIPTION
**Pull request by chenwany**
### Description of changes
* Merged feature branch to surface headnode bootstrap failures to develop branch
* The main changes are:
  * During the bootstrap of the head node, if any of custom scripts (OnNodeStart or OnNodeConfigured) fails, a generic error message will be sent to CloudFormation stack without providing specific reasons of the failure. Here we write error messages into a dedicated file, which will be extracted and sent to CloudFormation stack.
  *  During the bootstrap of the head node, if any of chef recipes fails, a generic error message will be sent to CloudFormation stack without providing specific reasons of the failure. Here we create a chef handler and apply to recipes. The handler will write error messages into a dedicated file, which will be extracted and sent to CloudFormation stack.
* Included PRs:
  * https://github.com/aws/aws-parallelcluster-cookbook/pull/1595
  * https://github.com/aws/aws-parallelcluster-cookbook/pull/1596
  * https://github.com/aws/aws-parallelcluster-cookbook/pull/1609
  * https://github.com/aws/aws-parallelcluster-cookbook/pull/1632
  * https://github.com/aws/aws-parallelcluster-cookbook/pull/1641
### Tests
* Test through manually tests and integration test: https://github.com/aws/aws-parallelcluster/pull/4830
* Describe the added/modified tests.

### References
* Link to related PRs in other packages (i.e. cookbook, node).
* Link to documentation useful to understand the changes.

### Checklist
- [ ] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [ ] Check all commits' messages are clear, describing what and why vs how.
- [ ] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [ ] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: cookbooks/aws-parallelcluster-config/templates/default/init/fetch_and_run.erb [hunk 1]
```diff
@@ -21,22 +21,40 @@ function error_exit () {
...
 function download_run (){
   url=$1
   shift
   scheme=$(echo "${url}"| cut -d: -f1)
+  url_processed="$(process_url "$url")"
   tmpfile=$(mktemp)
   trap "/bin/rm -f $tmpfile" RETURN
   if [ "${scheme}" == "s3" ]; then
-    <%= node['cluster']['cookbook_virtualenv_path'] %>/bin/aws --region ${cfn_region} s3 cp ${url} - > $tmpfile || return 1
+    <%= node['cluster']['cookbook_virtualenv_path'] %>/bin/aws --region ${cfn_region} s3 cp ${url} - > $tmpfile || error_exit "Failed to download ${custom_script} script ${url_processed} using aws s3, return code: $?."
```
Comment (davprat) [inline cookbooks/aws-parallelcluster-config/templates/default/init/fetch_and_run.erb:52]: [Question] Is inclusion of the URL something we need to address for security?

File: cookbooks/aws-parallelcluster-config/templates/default/init/fetch_and_run.erb [hunk 2]
```diff
@@ -21,22 +21,40 @@ function error_exit () {
...
   else
-    wget -qO- ${url} > $tmpfile || return 1
+    wget -qO- ${url} > $tmpfile || error_exit "Failed to download ${custom_script} script ${url_processed} using wget, return code: $?."
   fi
-  chmod +x $tmpfile || return 1
-  $tmpfile "$@" || error_exit "Failed to run ${ACTION}, ${url} failed with non 0 return code: $?"
+  chmod +x $tmpfile || error_exit "Failed to run ${custom_script} script due to a failure in making the file executable, return code: $?."
+  $tmpfile "$@" || error_exit "Failed to execute ${custom_script} script ${url_processed}, return code: $?."
```
Comment (davprat) [inline cookbooks/aws-parallelcluster-config/templates/default/init/fetch_and_run.erb:57]: Same question as above.

**Review by davprat — DISMISSED** [review]
LGTM

File: cookbooks/aws-parallelcluster-config/templates/default/init/fetch_and_run.erb (hunk 1 above)
Comment (chenwany) [reply cookbooks/aws-parallelcluster-config/templates/default/init/fetch_and_run.erb:52]: This PR merge the existing comments from feature branch to the develop branch. I will create another PR to address the security.

**Review by davprat — APPROVED** [review]

**Pull request merged**